Implement openstack v2 glance image api

vdloo opened the pull request:
## Use Openstack Glance v2 API for image manipulation

Previously all image functionality was available through the
compute API. This deprecated proxied API does not offer all
functionality that the Glance Image service API offers.
See https://developer.openstack.org/api-ref/compute/
> These APIs are proxy calls to the Image service. Nova has deprecated all
> the proxy APIs and users should use the native APIs instead. These will fail
> with a 404 starting from microversion 2.36. See: Relevant Image APIs.
For example, managing image visibility and sharing machine images across
tenants can not be done using the proxied image API in the compute endpoint,
but it can be done with the Glance Image API.
See https://developer.openstack.org/api-ref/image/v2/index.html#list-image-members

### Status

Haven't done real world testing with this branch yet, also I want to implement some more ex_ methods to help facilitate image sharing. But before I do that I'm curious what you guys think about the OpenStackImageConnection and the double init call to get the api client for the glance api
```
self.connectionCls = self.image_connectionCls
super(OpenStack_2_NodeDriver, self).__init__(*args, **kwargs)
```

vdloo committed:
implement openstack v2 glance image api

currently the compute driver uses the nova proxy to the glance image api, but this proxied api is deprecated and can not handle functionality like image sharing (setting an image to shared, creating new image members and accepting images as image members). this commit adds a connection to the glance API in the v2 driver

vdloo committed:
use glance api for openstack v2 list_images

vdloo committed:
add glance endpoint to openstack auth fixture

vdloo committed:
implement openstack ex_update_image

so visibility can be set on images

vdloo committed:
openstack_identity tests check for glance

pquentin commented:
I'm sorry that nobody has looked at this yet, we're lacking reviewers! Something that could help would be to rebase your branch on top of trunk: this would show coverage data. Please consider reviewing other pull requests too. Thanks.

pquentin commented on libcloud/compute/drivers/openstack.py:
```diff
@@ -2446,22 +2466,108 @@ def encode_data(self, data):
+        """
+        Get a NodeImage using the V2 Glance API
+
+        @inherits: :class:`NodeDriver.get_image`
```
This actually inherits get_image from OpenStack_1_1_NodeDriver.

pquentin commented on libcloud/compute/drivers/openstack.py:
```diff
@@ -2446,22 +2466,108 @@ def encode_data(self, data):
+        Lists all active images using the V2 Glance API
+
+        @inherits: :class:`NodeDriver.list_images`
+        """
```
Can you please document location and ex_only_active?

pquentin commented on libcloud/compute/drivers/openstack.py:
```diff
@@ -1283,19 +1290,32 @@ def _to_images(self, obj, ex_only_active):
+
+        if min_ram is None:
+            min_ram = api_image.get('minRam')
+        min_disk = api_image.get('min_disk')
```
nitpick (you don't need to change that): it's surprising to visually group min_ram and min_disk. I guess I would have added newlines *after* the if statements.

pquentin reviewed (changes requested):
Thanks! This looks good! Only two docstrings to update before I can merge this.

vdloo committed:
add space after if min_ram

>  it's surprising to visually group min_ram and min_disk. I guess I
>  would have added newlines after the if statements. - pquentin

vdloo committed:
fix OpenStack_2_NodeDriver docstring

> This actually inherits get_image from
> OpenStack_1_1_NodeDriver. - pquentin

vdloo committed:
docstr location and ex_only_active OS list_images

vdloo committed:
fix line too long linting errors in OS driver

fixes:
libcloud/compute/drivers/openstack.py:2488:80: E501 line too long (80 > 79 characters)
libcloud/compute/drivers/openstack.py:2489:80: E501 line too long (84 > 79 characters)
libcloud/compute/drivers/openstack.py:2492:80: E501 line too long (82 > 79 characters)
libcloud/compute/drivers/openstack.py:2494:80: E501 line too long (88 > 79 characters)

vdloo committed:
use image_connection in v2 OS update_image

vdloo commented:
OpenStack_1_1_NodeDriver related methods before this change:

```
In [6]: [i for i in conn.list_images() if 'ubuntu' in i.name][1]
Out[6]: <NodeImage: id=ba545873-27f2-4d55-a33c-147bcab677da, name=ubuntu-server-xenial-lts-64-bit, driver=OpenStack  ...>

In [9]: image = [i for i in conn.list_images() if 'ubuntu' in i.name][1]

In [10]: dir(image)
Out[10]: 
['__class__',
 '__delattr__',
 '__dict__',
 '__doc__',
 '__format__',
 '__getattribute__',
 '__hash__',
 '__init__',
 '__module__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_uuid',
 'driver',
 'extra',
 'get_uuid',
 'id',
 'name',
 'uuid']

In [11]: image.extra
Out[11]: 
{'created': u'2018-01-23T14:20:10Z',
 'metadata': {u'os_distro': u'ubuntu', u'os_type': u'linux'},
 'minDisk': 3,
 'minRam': 0,
 'progress': 100,
 'serverId': None,
 'status': u'ACTIVE',
 'updated': u'2018-01-23T14:34:36Z'}

In [13]: conn.get_image(image.id)
Out[13]: <NodeImage: id=ba545873-27f2-4d55-a33c-147bcab677da, name=ubuntu-server-xenial-lts-64-bit, driver=OpenStack  ...>
```

OpenStack_1_1_NodeDriver related methods after this change:

```
In [3]: [i for i in conn.list_images() if 'ubuntu' in i.name][1]
Out[3]: <NodeImage: id=ba545873-27f2-4d55-a33c-147bcab677da, name=ubuntu-server-xenial-lts-64-bit, driver=OpenStack  ...>

In [4]: image = [i for i in conn.list_images() if 'ubuntu' in i.name][1]

In [5]: dir(image)
Out[5]: 
['__class__',
 '__delattr__',
 '__dict__',
 '__doc__',
 '__format__',
 '__getattribute__',
 '__hash__',
 '__init__',
 '__module__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_uuid',
 'driver',
 'extra',
 'get_uuid',
 'id',
 'name',
 'uuid']

In [6]: image.extra
Out[6]: 
{'created': u'2018-01-23T14:20:10Z',
 'metadata': {u'os_distro': u'ubuntu', u'os_type': u'linux'},
 'minDisk': 3,
 'minRam': 0,
 'os_type': None,
 'progress': 100,
 'serverId': None,
 'status': u'ACTIVE',
 'updated': u'2018-01-23T14:34:36Z',
 'visibility': None}

In [7]: conn.get_image(image.id)
Out[7]: <NodeImage: id=ba545873-27f2-4d55-a33c-147bcab677da, name=ubuntu-server-xenial-lts-64-bit, driver=OpenStack  ...>
```

New OpenStack_2_NodeDriver related methods:

```
In [3]: [i for i in conn.list_images() if 'ubuntu' in i.name][1]
Out[3]: <NodeImage: id=ba545873-27f2-4d55-a33c-147bcab677da, name=ubuntu-server-xenial-lts-64-bit, driver=OpenStack  ...>

In [4]: image = [i for i in conn.list_images() if 'ubuntu' in i.name][1]

In [6]: image.extra
Out[6]: 
{'created': u'2018-01-23T14:20:10Z',
 'metadata': None,
 'minDisk': 3,
 'minRam': 0,
 'os_type': u'linux',
 'progress': None,
 'serverId': None,
 'status': u'active',
 'updated': u'2018-01-23T14:34:36Z',
 'visibility': u'public'}

In [7]: conn.get_image(image.id)
Out[7]: <NodeImage: id=ba545873-27f2-4d55-a33c-147bcab677da, name=ubuntu-server-xenial-lts-64-bit, driver=OpenStack  ...>
```

Looks like metadata as before is gone. The os_type is present at the top level, but os_distro is not in the metadata anymore but could also become a top level key like os_type. Haven't implemented that in this PR but maybe that's something for the future. api response looks like this:
```
{u'checksum': u'eb470662a94e5a17fa00bc0e3a674018',
 u'container_format': u'bare',
 u'created_at': u'2018-01-23T14:21:36Z',
 u'direct_url': u'file:///var/lib/glance/images/1a67daa3-0dc4-448e-9fa9-ff6cc03ea622',
 u'disk_format': u'qcow2',
 u'file': u'/v2/images/1a67daa3-0dc4-448e-9fa9-ff6cc03ea622/file',
 u'id': u'1a67daa3-0dc4-448e-9fa9-ff6cc03ea622',
 u'locations': [{u'metadata': {u'id': u'bla',
                               u'mountpoint': u'/var/lib/glance/images',
                               u'share_location': u'nfs://1.2.3.4/bla',
                               u'type': u'nfs'},
                 u'url': u'file:///var/lib/glance/images/1a67daa3-0dc4-448e-9fa9-ff6cc03ea622'}],
 u'min_disk': 3,
 u'min_ram': 0,
 u'name': u'ubuntu-server-trusty-lts-64-bit',
 u'os_distro': u'ubuntu',
 u'os_type': u'linux',
 u'owner': u'1bcce578589e46b4b0221f11d0fbe042',
 u'protected': True,
 u'schema': u'/v2/schemas/image',
 u'self': u'/v2/images/1a67daa3-0dc4-448e-9fa9-ff6cc03ea622',
 u'size': 262341120,
 u'status': u'active',
 u'tags': [],
 u'updated_at': u'2018-01-23T14:34:31Z',
 u'virtual_size': None,
 u'visibility': u'public'}
```

And setting visibility with ex_update_image:
```
In [6]: private_image.extra['visibility']
Out[6]: u'private'

In [7]: conn.ex_update_image(private_image.id, [{'op': 'replace', 'path': '/visibility', 'value': 'shared'}])
Out[7]: <NodeImage: id=df0a0e9b-c524-4d48-a209-c04ebc92aabb, name=base-precise-20161101, driver=OpenStack  ...>

In [9]: conn.get_image(private_image.id).extra['visibility']
Out[9]: u'shared'
```

vdloo commented:
@pquentin this PR should be good to go

vdloo committed:
add OpenStack_2_NodeDriver to api docs

vdloo commented:
ping @pquentin

pquentin commented:
Thanks @vdloo! Finally merged into trunk. ✨
